File: 1023-time-based-key-value-store/time-based-key-value-store.java
class TimeMap {

    class Pair {
        String value;
        int time;

        public Pair(String value, int time) {
            this.value = value;
            this.time = time;
        }
    }

    Map<String, List<Pair>> map;

    public TimeMap() {
        map = new HashMap<>();
    }

    public void set(String key, String value, int timestamp) {
        map.computeIfAbsent(key, k -> new ArrayList<>()).add(new Pair(value, timestamp));
    }

    public String get(String key, int timestamp) {
        if (!map.containsKey(key))
            return "";

        List<Pair> values = map.get(key);

        int low = 0;
        int high = values.size() - 1;
        String res = "";

        while (low <= high) {
            int mid = low + ((high - low) >> 1);

            int time = values.get(mid).time;

            if (time <= timestamp) {
                res = values.get(mid).value;
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return res;
    }
}

File: 1023-time-based-key-value-store/time-based-key-value-store.py
class TimeMap:

    def __init__(self):
        self.m = {}
        

    def set(self, key: str, value: str, timestamp: int) -> None:
        if key not in self.m:
            self.m[key] = [(value, timestamp)]
        else:
            self.m[key].append((value, timestamp))
        

    def get(self, key: str, timestamp: int) -> str:
        if key not in self.m:
            return ""
        
        lst = self.m[key]

        l = 0
        r = len(lst) - 1

        while l <= r:
            mid = l + (r-l)//2

            value, time = lst[mid]

            if time == timestamp:
                return value
            elif time < timestamp:
                l = mid + 1
            else:
                r = mid - 1
        
        return "" if r < 0 else lst[r][0]
        


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)